File: gm_menubar.swift
import Cocoa

// ============================================================
// GM Display — menu-bar app (⚔️)
//
// This app is THE thing that runs the server. Nothing else should:
//
//   * It runs server.py straight out of the vault — never a copy baked into
//     the app bundle — so a restart always runs the code that is on disk now.
//   * The server is its child process, so macOS asks once whether "GM
//     Display" may read your Documents folder and remembers the answer. (A
//     launchd job running python on its own gets "Operation not permitted" on
//     everything in ~/Documents, which is why the old daemon never worked.)
//   * It owns port 7680: starting or restarting takes the port from whoever
//     holds it, and it unloads the old launchd job so that cannot respawn.
//   * The server also restarts ITSELF when its own .py files change, so an
//     update is live without touching this menu at all. "Restart Server" is
//     for when you want to be sure.
//
// Settings live beside server.py in gm-display.conf (read on every start).
// Log: ~/Library/Logs/gm-display.log
// ============================================================

let PORT = 7680
let LEGACY_LABEL = "com.gm-display.server"

struct Config {
    var python = "/usr/bin/python3"
    var browser = ""               // e.g. open -na Dia --args --profile-directory=Default {url}
    var openAtLaunch = true
    var serverArgs: [String] = []
}

final class AppDelegate: NSObject, NSApplicationDelegate {
    var statusItem: NSStatusItem!
    var statusLine: NSMenuItem!
    var detailLine: NSMenuItem!
    var startStopItem: NSMenuItem!

    var server: Process?
    var stopping = false           // an exit we asked for — do not restart
    var recentExits: [Date] = []
    var healthTimer: Timer?
    var pendingURLs: [String] = []
    var retiredPIDs = Set<Int32>()  // servers we stopped on purpose

    // Where server.py lives. build_app.sh writes it into Info.plist.
    lazy var toolsDir: String = {
        if let d = Bundle.main.object(forInfoDictionaryKey: "GMDToolsDir") as? String, !d.isEmpty {
            return d
        }
        return NSHomeDirectory() + "/Documents/RPG/Campaign Vault/.tools/gm-display"
    }()
    var serverScript: String { toolsDir + "/server.py" }
    var confPath: String { toolsDir + "/gm-display.conf" }
    var logPath: String { NSHomeDirectory() + "/Library/Logs/gm-display.log" }

    // MARK: - lifecycle

    func applicationDidFinishLaunching(_ notification: Notification) {
        NSApp.setActivationPolicy(.accessory)
        statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)
        statusItem.button?.title = "⚔️"
        buildMenu()

        let cfg = readConfig()
        startServer(openPageWhenUp: cfg.openAtLaunch)

        healthTimer = Timer.scheduledTimer(withTimeInterval: 3.0, repeats: true) { [weak self] _ in
            self?.checkHealth()
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self] in self?.checkHealth() }
    }

    func applicationWillTerminate(_ notification: Notification) {
        stopServerSync()
    }

    // MARK: - menu

    func buildMenu() {
        let menu = NSMenu()
        statusLine = NSMenuItem(title: "Server: starting…", action: nil, keyEquivalent: "")
        statusLine.isEnabled = false
        menu.addItem(statusLine)
        detailLine = NSMenuItem(title: "", action: nil, keyEquivalent: "")
        detailLine.isEnabled = false
        menu.addItem(detailLine)
        menu.addItem(NSMenuItem.separator())

        add(menu, "Open GM Page", #selector(openGMPage), "g")
        add(menu, "Open Projector", #selector(openProjector), "")
        add(menu, "Open Sidecar", #selector(openSidecar), "")
        add(menu, "Preview Player View", #selector(openPlayerPreview), "p")
        menu.addItem(NSMenuItem.separator())

        add(menu, "Restart Server", #selector(restartServer), "r")
        startStopItem = add(menu, "Stop Server", #selector(toggleServer), "")
        add(menu, "Show Log", #selector(showLog), "l")
        add(menu, "Edit Settings (gm-display.conf)", #selector(editConfig), "")
        add(menu, "Reveal Server Folder", #selector(revealFolder), "")
        menu.addItem(NSMenuItem.separator())

        add(menu, "Quit GM Display", #selector(quitApp), "q")
        statusItem.menu = menu
    }

    @discardableResult
    func add(_ menu: NSMenu, _ title: String, _ action: Selector, _ key: String) -> NSMenuItem {
        let item = NSMenuItem(title: title, action: action, keyEquivalent: key)
        item.target = self
        menu.addItem(item)
        return item
    }

    // MARK: - config

    func readConfig() -> Config {
        var cfg = Config()
        guard let text = try? String(contentsOfFile: confPath, encoding: .utf8) else { return cfg }
        for raw in text.components(separatedBy: .newlines) {
            let line = raw.trimmingCharacters(in: .whitespaces)
            if line.isEmpty || line.hasPrefix("#") { continue }
            if line.hasPrefix("--") { cfg.serverArgs.append(line); continue }
            guard let colon = line.firstIndex(of: ":") else { continue }
            let key = line[..<colon].trimmingCharacters(in: .whitespaces).lowercased()
            let value = line[line.index(after: colon)...].trimmingCharacters(in: .whitespaces)
            switch key {
            case "python": if !value.isEmpty { cfg.python = value }
            case "browser": cfg.browser = value
            case "open-gm-page-at-launch": cfg.openAtLaunch = !["no", "false", "0", "off"].contains(value.lowercased())
            default: break
            }
        }
        return cfg
    }

    // MARK: - server

    func startServer(openPageWhenUp: Bool = false) {
        stopping = false
        let cfg = readConfig()
        guard FileManager.default.fileExists(atPath: serverScript) else {
            setStatus("Server: server.py not found", icon: "⚔️ ✗")
            detailLine.title = serverScript
            return
        }
        retireLegacyDaemon()
        freePort()
        rotateLogIfBig()

        let fm = FileManager.default
        if !fm.fileExists(atPath: logPath) { fm.createFile(atPath: logPath, contents: nil) }
        let log = FileHandle(forWritingAtPath: logPath)
        log?.seekToEndOfFile()
        let stamp = ISO8601DateFormatter().string(from: Date())
        log?.write("\n==== GM Display menu-bar app starting server \(stamp) ====\n".data(using: .utf8)!)

        let p = Process()
        p.executableURL = URL(fileURLWithPath: cfg.python)
        p.arguments = [serverScript, "--no-browser"] + cfg.serverArgs
        p.currentDirectoryURL = URL(fileURLWithPath: toolsDir)
        var env = ProcessInfo.processInfo.environment
        env["GMD_SUPERVISOR"] = "tray"
        env["PYTHONUNBUFFERED"] = "1"
        p.environment = env
        if let log = log {
            p.standardOutput = log
            p.standardError = log
        }
        p.terminationHandler = { [weak self] proc in
            DispatchQueue.main.async { self?.serverExited(proc) }
        }
        do {
            try p.run()
            server = p
            setStatus("Server: starting…", icon: "⚔️ …")
            startStopItem.title = "Stop Server"
            if openPageWhenUp { whenUp { [weak self] in self?.openGMPage() } }
            whenUp { [weak self] in self?.flushPendingURLs() }
        } catch {
            setStatus("Server: could not start python", icon: "⚔️ ✗")
            detailLine.title = "\(cfg.python): \(error.localizedDescription)"
        }
    }

    func serverExited(_ proc: Process) {
        if retiredPIDs.remove(proc.processIdentifier) != nil { return }
        if server === proc { server = nil }
        if stopping { return }
        // Something else may have taken the port on purpose (server.py --force
        // from a Terminal). Do not fight it for the port.
        fetchHealth { [weak self] h in
            guard let self = self else { return }
            if let h = h, (h["supervisor"] as? String) != "tray" {
                self.setStatus("Server: running from Terminal (not managed here)", icon: "⚔️ ⌨")
                self.startStopItem.title = "Take Over Server"
                return
            }
            let now = Date()
            self.recentExits = self.recentExits.filter { now.timeIntervalSince($0) < 60 } + [now]
            if self.recentExits.count >= 4 {
                self.setStatus("Server: keeps crashing — see Show Log", icon: "⚔️ ✗")
                self.startStopItem.title = "Start Server"
                return
            }
            self.setStatus("Server: exited — restarting…", icon: "⚔️ …")
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) { [weak self] in
                guard let self = self, !self.stopping, self.server == nil else { return }
                self.startServer()
            }
        }
    }

    // Stop our server and wait for it (SIGTERM, then SIGKILL after 4s).
    func stopServerSync() {
        stopping = true
        guard let p = server, p.isRunning else { server = nil; return }
        let pid = p.processIdentifier
        DispatchQueue.main.async { self.retiredPIDs.insert(pid) }
        p.terminate()
        let deadline = Date().addingTimeInterval(4)
        while p.isRunning && Date() < deadline { usleep(100_000) }
        if p.isRunning { kill(p.processIdentifier, SIGKILL) }
        server = nil
    }

    // The port has one owner. Whatever is listening on it now — a Terminal
    // copy, a leftover from an old install — is asked to stop, then made to.
    func freePort() {
        let out = sh("/usr/sbin/lsof -nP -ti tcp:\(PORT) -sTCP:LISTEN 2>/dev/null")
        let mine = server?.processIdentifier ?? -1
        let pids = out.split(separator: "\n").compactMap { Int32($0.trimmingCharacters(in: .whitespaces)) }
            .filter { $0 != mine && $0 != getpid() }
        if pids.isEmpty { return }
        for pid in pids { kill(pid, SIGTERM) }
        let deadline = Date().addingTimeInterval(3)
        while Date() < deadline && pids.contains(where: { kill($0, 0) == 0 }) { usleep(100_000) }
        for pid in pids where kill(pid, 0) == 0 { kill(pid, SIGKILL) }
        usleep(300_000)
    }

    // The old launchd daemon ran python without Documents access and fought
    // this app for the port (KeepAlive respawned it thousands of times). Make
    // sure it is gone.
    func retireLegacyDaemon() {
        let uid = getuid()
        _ = sh("/bin/launchctl bootout gui/\(uid)/\(LEGACY_LABEL) 2>/dev/null")
        let plist = NSHomeDirectory() + "/Library/LaunchAgents/\(LEGACY_LABEL).plist"
        if FileManager.default.fileExists(atPath: plist) {
            try? FileManager.default.trashItem(at: URL(fileURLWithPath: plist), resultingItemURL: nil)
        }
    }

    func rotateLogIfBig() {
        let fm = FileManager.default
        guard let attrs = try? fm.attributesOfItem(atPath: logPath),
              let size = attrs[.size] as? NSNumber, size.intValue > 5_000_000 else { return }
        let old = logPath + ".old"
        try? fm.removeItem(atPath: old)
        try? fm.moveItem(atPath: logPath, toPath: old)
    }

    // MARK: - health

    func fetchHealth(_ done: @escaping ([String: Any]?) -> Void) {
        var req = URLRequest(url: URL(string: "http://127.0.0.1:\(PORT)/api/health")!)
        req.timeoutInterval = 1.5
        URLSession.shared.dataTask(with: req) { data, resp, _ in
            var out: [String: Any]? = nil
            if let data = data, let http = resp as? HTTPURLResponse, http.statusCode == 200 {
                out = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
            }
            DispatchQueue.main.async { done(out) }
        }.resume()
    }

    // Run `then` once the server answers (gives up after ~20s).
    func whenUp(tries: Int = 40, _ then: @escaping () -> Void) {
        fetchHealth { [weak self] h in
            if h != nil { then(); return }
            if tries <= 0 { return }
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { self?.whenUp(tries: tries - 1, then) }
        }
    }

    func checkHealth() {
        fetchHealth { [weak self] h in
            guard let self = self else { return }
            guard let h = h else {
                if self.server?.isRunning == true {
                    self.setStatus("Server: starting…", icon: "⚔️ …")
                } else if !self.stopping && self.recentExits.count < 4 {
                    self.setStatus("Server: not running", icon: "⚔️ ⚠")
                } else if self.stopping {
                    self.setStatus("Server: stopped", icon: "⚔️ ◌")
                    self.startStopItem.title = "Start Server"
                }
                return
            }
            let pid = (h["pid"] as? NSNumber)?.intValue ?? 0
            let managed = (h["supervisor"] as? String) == "tray"
            var since = ""
            if let t = h["started"] as? Double {
                let f = DateFormatter(); f.dateFormat = "HH:mm:ss"
                since = " · since \(f.string(from: Date(timeIntervalSince1970: t)))"
            }
            if managed {
                self.setStatus("Server: running ✓", icon: "⚔️")
                self.startStopItem.title = "Stop Server"
            } else {
                self.setStatus("Server: running from Terminal (not managed here)", icon: "⚔️ ⌨")
                self.startStopItem.title = "Take Over Server"
            }
            let py = (h["python"] as? String).map { " · python \($0)" } ?? ""
            self.detailLine.title = "pid \(pid)\(since)\(py) · port \(PORT)"
        }
    }

    func setStatus(_ text: String, icon: String) {
        statusLine.title = text
        statusItem.button?.title = icon
    }

    // MARK: - actions

    @objc func restartServer() {
        setStatus("Server: restarting…", icon: "⚔️ …")
        recentExits = []
        DispatchQueue.global().async { [weak self] in
            self?.stopServerSync()
            DispatchQueue.main.async { self?.startServer() }
        }
    }

    @objc func toggleServer() {
        if startStopItem.title == "Stop Server" {
            setStatus("Server: stopping…", icon: "⚔️ …")
            DispatchQueue.global().async { [weak self] in
                self?.stopServerSync()
                DispatchQueue.main.async {
                    self?.setStatus("Server: stopped", icon: "⚔️ ◌")
                    self?.startStopItem.title = "Start Server"
                }
            }
        } else {
            recentExits = []
            startServer()
        }
    }

    @objc func openGMPage() { openInBrowser("/gm.html") }
    @objc func openProjector() { openInBrowser("/display.html?display=map") }
    @objc func openSidecar() { openInBrowser("/display.html?display=show") }
    @objc func openPlayerPreview() { openInBrowser("/remote.html?preview=1") }

    // Opens in the browser (and PROFILE) named in gm-display.conf, because the
    // whole app lives in that profile's localStorage.
    func openInBrowser(_ path: String) {
        let url = "http://localhost:\(PORT)\(path)"
        let cfg = readConfig()
        if cfg.browser.isEmpty {
            NSWorkspace.shared.open(URL(string: url)!)
            return
        }
        let quoted = "'" + url + "'"
        let cmd = cfg.browser.contains("{url}")
            ? cfg.browser.replacingOccurrences(of: "{url}", with: quoted)
            : cfg.browser + " " + quoted
        DispatchQueue.global().async { _ = self.sh(cmd) }
    }

    @objc func showLog() {
        _ = sh("/usr/bin/open -a Console '\(logPath)'")
    }

    @objc func editConfig() {
        if !FileManager.default.fileExists(atPath: confPath) {
            FileManager.default.createFile(atPath: confPath, contents: "# See gm-display.conf in the repo\n".data(using: .utf8))
        }
        NSWorkspace.shared.open(URL(fileURLWithPath: confPath))
    }

    @objc func revealFolder() {
        NSWorkspace.shared.activateFileViewerSelecting([URL(fileURLWithPath: serverScript)])
    }

    @objc func quitApp() {
        stopServerSync()
        NSApp.terminate(nil)
    }

    // MARK: - gm:// links

    func application(_ application: NSApplication, open urls: [URL]) {
        for url in urls where url.scheme == "gm" {
            if url.host == "control" {
                switch url.path {
                case "/restart": restartServer()
                case "/stop": if startStopItem.title == "Stop Server" { toggleServer() }
                case "/start": if server == nil { recentExits = []; startServer() }
                default: break
                }
                continue
            }
            pendingURLs.append(url.absoluteString)
        }
        flushPendingURLs()
    }

    func flushPendingURLs() {
        if pendingURLs.isEmpty { return }
        whenUp { [weak self] in
            guard let self = self else { return }
            let urls = self.pendingURLs
            self.pendingURLs = []
            for u in urls { self.postCommand(u) }
        }
    }

    // The server parses gm:// URLs itself ({"url": "gm://map/..."}).
    func postCommand(_ gmURL: String) {
        guard let body = try? JSONSerialization.data(withJSONObject: ["url": gmURL]) else { return }
        var req = URLRequest(url: URL(string: "http://127.0.0.1:\(PORT)/api/command")!)
        req.httpMethod = "POST"
        req.httpBody = body
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        URLSession.shared.dataTask(with: req) { _, _, error in
            if let error = error { NSLog("gm:// post failed: \(error)") }
        }.resume()
    }

    // MARK: - helpers

    @discardableResult
    func sh(_ command: String) -> String {
        let p = Process()
        p.executableURL = URL(fileURLWithPath: "/bin/sh")
        p.arguments = ["-c", command]
        let pipe = Pipe()
        p.standardOutput = pipe
        p.standardError = FileHandle.nullDevice
        do { try p.run() } catch { return "" }
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        p.waitUntilExit()
        return String(data: data, encoding: .utf8) ?? ""
    }
}

// Only one of us. A second launch (Login Items plus a manual open, say) hands
// over to the first and quits, rather than starting a second server.
let me = Bundle.main.bundleIdentifier ?? "com.gm-display.app"
let others = NSRunningApplication.runningApplications(withBundleIdentifier: me)
    .filter { $0.processIdentifier != getpid() }
if !others.isEmpty {
    others.first?.activate(options: [])
    exit(0)
}

let app = NSApplication.shared
let delegate = AppDelegate()
app.delegate = delegate
app.run()

File: gm_menubar.py
#!/usr/bin/env python3
"""
GM Display Menu Bar App
Sits in the macOS menu bar (⚔️), runs the HTTP server, handles gm:// URLs.

Architecture:
  - NSApplication with NSStatusItem (menu bar icon)
  - HTTP server runs in a background thread
  - Handles gm:// URL scheme via Apple Events
  - No dock icon (accessory app)
"""

import os
import sys
import json
import threading
import time
import http.server
import urllib.parse
import urllib.request
import signal

# --- Import PyObjC ---
try:
    import objc
    from AppKit import (
        NSApplication, NSApp, NSObject, NSStatusBar, NSMenu, NSMenuItem,
        NSVariableStatusItemLength, NSWorkspace, NSImage,
        NSApplicationActivationPolicyAccessory, NSEventTrackingRunLoopMode
    )
    from Foundation import (
        NSURL, NSTimer, NSRunLoop, NSDefaultRunLoopMode, NSLog
    )
    HAS_PYOBJC = True
except ImportError:
    HAS_PYOBJC = False
    print("PyObjC not available — install with: pip3 install pyobjc-framework-Cocoa")
    print("Falling back to server-only mode.")


# ============================================================
# HTTP SERVER (same as server.py, embedded for single-process)
# ============================================================

PORT = 7680
IMAGE_ROOTS = []
pending_command = None
command_lock = threading.Lock()


class GMHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=STATIC_DIR, **kwargs)

    def do_GET(self):
        parsed = urllib.parse.urlparse(self.path)

        # Poll for pending commands
        if parsed.path == '/api/command':
            global pending_command
            with command_lock:
                cmd = pending_command
                pending_command = None
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps(cmd).encode() if cmd else b'null')
            return

        # Serve map/image files from configured roots
        if parsed.path.startswith('/maps/'):
            rel_path = urllib.parse.unquote(parsed.path[6:])
            for root in IMAGE_ROOTS:
                full_path = os.path.join(root, rel_path)
                if os.path.isfile(full_path):
                    try:
                        self.send_file(full_path)
                    except Exception as e:
                        print(f"Error serving {full_path}: {e}")
                        self.send_error(500, f"Error serving file: {e}")
                    return
            print(f"404: '{rel_path}' not found in {IMAGE_ROOTS}")
            self.send_error(404, f"Image not found: {rel_path}")
            return

        # List available maps
        if parsed.path == '/api/maps':
            maps = []
            for root in IMAGE_ROOTS:
                for dirpath, _, filenames in os.walk(root):
                    for f in filenames:
                        if f.lower().endswith(('.png', '.jpg', '.jpeg', '.webp', '.gif')):
                            rel = os.path.relpath(os.path.join(dirpath, f), root)
                            maps.append({'name': f, 'path': '/maps/' + urllib.parse.quote(rel)})
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps(maps).encode())
            return

        # Health check
        if parsed.path == '/api/health':
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'status': 'ok', 'roots': len(IMAGE_ROOTS)}).encode())
            return

        # Default: serve static files (no cache on HTML)
        if parsed.path == '/' or parsed.path == '':
            self.path = '/gm_display.html'

        if self.path.endswith('.html') or self.path.split('?')[0].endswith('.html'):
            clean_path = self.path.split('?')[0].lstrip('/')
            file_path = os.path.join(STATIC_DIR, clean_path)
            if os.path.isfile(file_path):
                with open(file_path, 'rb') as f:
                    data = f.read()
                self.send_response(200)
                self.send_header('Content-Type', 'text/html; charset=utf-8')
                self.send_header('Content-Length', len(data))
                self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
                self.send_header('Pragma', 'no-cache')
                self.end_headers()
                self.wfile.write(data)
                return

        super().do_GET()

    def do_POST(self):
        parsed = urllib.parse.urlparse(self.path)

        if parsed.path == '/api/command':
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length).decode() if content_length else ''

            params = urllib.parse.parse_qs(parsed.query)
            cmd = None
            if body:
                try:
                    cmd = json.loads(body)
                except json.JSONDecodeError:
                    params.update(urllib.parse.parse_qs(body))

            if not cmd:
                url = params.get('url', [''])[0]
                if url:
                    cmd = parse_gm_url(url)
                else:
                    cmd = {
                        'action': params.get('action', ['show'])[0],
                        'file': params.get('file', [''])[0]
                    }

            global pending_command
            with command_lock:
                pending_command = cmd

            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({'ok': True, 'command': cmd}).encode())
            return

        self.send_error(405)

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def send_file(self, path):
        ext = os.path.splitext(path)[1].lower()
        mime_types = {
            '.png': 'image/png', '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg',
            '.gif': 'image/gif', '.webp': 'image/webp',
        }
        mime = mime_types.get(ext, 'application/octet-stream')
        with open(path, 'rb') as f:
            data = f.read()
        self.send_response(200)
        self.send_header('Content-Type', mime)
        self.send_header('Content-Length', len(data))
        self.send_header('Cache-Control', 'max-age=3600')
        self.end_headers()
        self.wfile.write(data)

    def log_message(self, format, *args):
        if args and '404' in str(args[0]):
            super().log_message(format, *args)


def parse_gm_url(url):
    """Parse a gm://action/path URL into a command dict."""
    parsed = urllib.parse.urlparse(url)
    action = parsed.hostname  # 'show' or 'map'
    file_path = urllib.parse.unquote(parsed.path.lstrip('/'))
    return {
        'action': action or 'show',
        'file': f'/maps/{urllib.parse.quote(file_path)}'
    }


def send_command_to_server(url):
    """Send a gm:// URL to the running server via HTTP POST."""
    cmd = parse_gm_url(url)
    try:
        data = json.dumps(cmd).encode()
        req = urllib.request.Request(
            f'http://localhost:{PORT}/api/command',
            data=data,
            headers={'Content-Type': 'application/json'},
            method='POST'
        )
        urllib.request.urlopen(req, timeout=2)
        return True
    except Exception as e:
        print(f"Failed to send command: {e}")
        return False


def add_vault(vault_path):
    """Add a vault root and all its immediate subfolders to IMAGE_ROOTS."""
    if vault_path not in IMAGE_ROOTS:
        IMAGE_ROOTS.append(vault_path)
    try:
        for sub in os.listdir(vault_path):
            sub_path = os.path.join(vault_path, sub)
            if os.path.isdir(sub_path) and not sub.startswith('.') and sub_path not in IMAGE_ROOTS:
                IMAGE_ROOTS.append(sub_path)
    except PermissionError:
        print(f"Warning: cannot list {vault_path}")


def discover_image_roots():
    """Find vault and image directories."""
    # Primary: walk up from our own location
    script_dir = os.path.dirname(os.path.abspath(__file__))

    # If running from .app bundle, Resources/ contains our files
    # but the vault is at the original location
    if '/Contents/Resources' in script_dir:
        # Running from .app bundle — find vault via known paths
        home = os.path.expanduser('~')
        for name in ['Documents/RPG/Campaign Vault', 'Documents/Pathfinder', 'Documents/Obsidian Vault']:
            candidate = os.path.join(home, name)
            if os.path.isdir(candidate):
                add_vault(candidate)
    else:
        # Running from vault directly (.tools/gm-display/)
        vault = os.path.dirname(os.path.dirname(script_dir))
        if os.path.isdir(vault):
            add_vault(vault)

    if not IMAGE_ROOTS:
        # Fallback: scan ~/Documents for Obsidian vaults
        home = os.path.expanduser('~')
        docs = os.path.join(home, 'Documents')
        if os.path.isdir(docs):
            for entry in os.listdir(docs):
                candidate = os.path.join(docs, entry)
                if os.path.isdir(candidate) and (
                    os.path.isdir(os.path.join(candidate, '.obsidian')) or
                    os.path.isdir(os.path.join(candidate, 'Darkmoon Vale'))
                ):
                    add_vault(candidate)


def start_http_server():
    """Start the HTTP server (blocking — run in a thread)."""
    try:
        server = http.server.HTTPServer(('127.0.0.1', PORT), GMHandler)
        print(f"GM Display Server running on http://localhost:{PORT}")
        print(f"Image roots: {IMAGE_ROOTS}")
        server.serve_forever()
    except OSError as e:
        if 'Address already in use' in str(e):
            print(f"Port {PORT} already in use — another instance may be running")
        else:
            raise


# Resolve STATIC_DIR — find where gm_display.html lives
def find_static_dir():
    """Find the directory containing gm_display.html."""
    candidates = [
        os.path.dirname(os.path.abspath(__file__)),  # same dir as this script
    ]
    # If in .app bundle, also check Resources
    script_dir = os.path.dirname(os.path.abspath(__file__))
    if '/Contents/Resources' not in script_dir:
        # Not in bundle — also check .tools/gm-display/ relative to vault
        pass

    for d in candidates:
        if os.path.isfile(os.path.join(d, 'gm_display.html')):
            return d

    # Last resort
    return script_dir

STATIC_DIR = find_static_dir()


# ============================================================
# MENU BAR APP (PyObjC)
# ============================================================

if HAS_PYOBJC:

    class GMDisplayDelegate(NSObject):
        """NSApplication delegate — creates menu bar icon, manages server."""

        def applicationDidFinishLaunching_(self, notification):
            # Hide dock icon — we're a menu bar app only
            NSApp.setActivationPolicy_(NSApplicationActivationPolicyAccessory)

            # Create status bar item
            self.statusItem = NSStatusBar.systemStatusBar().statusItemWithLength_(
                NSVariableStatusItemLength
            )
            self.statusItem.setTitle_("\u2694\uFE0F")  # ⚔️
            self.statusItem.setHighlightMode_(True)

            self._server_ok = False
            self._build_menu()

            # Start server
            self._server_thread = threading.Thread(target=start_http_server, daemon=True)
            self._server_thread.start()

            # Health check every 5 seconds
            self._health_timer = NSTimer.scheduledTimerWithTimeInterval_target_selector_userInfo_repeats_(
                3.0, self, b"healthCheck:", None, True
            )
            # Initial check after 1s
            NSTimer.scheduledTimerWithTimeInterval_target_selector_userInfo_repeats_(
                1.0, self, b"healthCheck:", None, False
            )

        def _build_menu(self):
            menu = NSMenu.alloc().init()

            # Status line (disabled, just informational)
            self._status_item = NSMenuItem.alloc().initWithTitle_action_keyEquivalent_(
                "Server: starting...", None, ""
            )
            self._status_item.setEnabled_(False)
            menu.addItem_(self._status_item)

            self._roots_item = NSMenuItem.alloc().initWithTitle_action_keyEquivalent_(
                f"Image roots: {len(IMAGE_ROOTS)}", None, ""
            )
            self._roots_item.setEnabled_(False)
            menu.addItem_(self._roots_item)

            menu.addItem_(NSMenuItem.separatorItem())

            # Open pages
            item = NSMenuItem.alloc().initWithTitle_action_keyEquivalent_(
                "Open GM Page", b"openGMPage:", ""
            )
            menu.addItem_(item)

            item = NSMenuItem.alloc().initWithTitle_action_keyEquivalent_(
                "Open Map Display (Projector)", b"openMapDisplay:", ""
            )
            menu.addItem_(item)

            item = NSMenuItem.alloc().initWithTitle_action_keyEquivalent_(
                "Open Image Display (Sidecar)", b"openSidecarDisplay:", ""
            )
            menu.addItem_(item)

            menu.addItem_(NSMenuItem.separatorItem())

            # Restart
            item = NSMenuItem.alloc().initWithTitle_action_keyEquivalent_(
                "Restart Server", b"restartServer:", ""
            )
            menu.addItem_(item)

            menu.addItem_(NSMenuItem.separatorItem())

            # Quit
            item = NSMenuItem.alloc().initWithTitle_action_keyEquivalent_(
                "Quit GM Display", b"quitApp:", "q"
            )
            menu.addItem_(item)

            self.statusItem.setMenu_(menu)

        @objc.typedSelector(b"v@:@")
        def healthCheck_(self, timer):
            """Periodic server health check — updates menu bar icon."""
            try:
                req = urllib.request.Request(f'http://localhost:{PORT}/api/health', method='GET')
                resp = urllib.request.urlopen(req, timeout=1)
                data = json.loads(resp.read())
                self._server_ok = True
                self._status_item.setTitle_(f"Server: running \u2714")
                self._roots_item.setTitle_(f"Image roots: {data.get('roots', '?')}")
                self.statusItem.setTitle_("\u2694\uFE0F")
            except Exception:
                self._server_ok = False
                self._status_item.setTitle_("Server: not responding \u2718")
                self.statusItem.setTitle_("\u2694\uFE0F \u26A0")

        @objc.typedSelector(b"v@:@")
        def openGMPage_(self, sender):
            NSWorkspace.sharedWorkspace().openURL_(
                NSURL.URLWithString_("http://localhost:7680")
            )

        @objc.typedSelector(b"v@:@")
        def openMapDisplay_(self, sender):
            NSWorkspace.sharedWorkspace().openURL_(
                NSURL.URLWithString_("http://localhost:7680/gm_display.html?mode=player&display=map")
            )

        @objc.typedSelector(b"v@:@")
        def openSidecarDisplay_(self, sender):
            NSWorkspace.sharedWorkspace().openURL_(
                NSURL.URLWithString_("http://localhost:7680/gm_display.html?mode=player&display=show")
            )

        @objc.typedSelector(b"v@:@")
        def restartServer_(self, sender):
            self._status_item.setTitle_("Server: restarting...")
            self.statusItem.setTitle_("\u2694\uFE0F ...")
            # Kill the old server thread (it's daemon, will die)
            # Start a new one
            import socket
            try:
                s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                s.connect(('127.0.0.1', PORT))
                s.close()
            except:
                pass
            self._server_thread = threading.Thread(target=start_http_server, daemon=True)
            self._server_thread.start()

        @objc.typedSelector(b"v@:@")
        def quitApp_(self, sender):
            NSApp.terminate_(None)

        # --- Handle gm:// URLs sent to the app ---
        def application_openURLs_(self, app, urls):
            for url in urls:
                url_string = str(url.absoluteString())
                if url_string.startswith('gm://'):
                    print(f"Received URL: {url_string}")
                    cmd = parse_gm_url(url_string)
                    global pending_command
                    with command_lock:
                        pending_command = cmd


# ============================================================
# MAIN
# ============================================================

def main():
    # Handle CLI mode: if called with a gm:// URL, POST it and exit
    for arg in sys.argv[1:]:
        if arg.startswith('gm://'):
            send_command_to_server(arg)
            return
        elif os.path.isdir(arg):
            IMAGE_ROOTS.append(os.path.abspath(arg))

    # Discover image roots
    discover_image_roots()

    if HAS_PYOBJC:
        # Run as menu bar app
        app = NSApplication.sharedApplication()
        delegate = GMDisplayDelegate.alloc().init()
        app.setDelegate_(delegate)
        app.run()
    else:
        # Fallback: just run the server
        print("Running in server-only mode (no menu bar icon)")
        import webbrowser
        threading.Timer(0.5, lambda: webbrowser.open(f'http://localhost:{PORT}')).start()
        start_http_server()


if __name__ == '__main__':
    main()
